File: sdks/node/src/multiUserClient.ts
import { randomUUID } from "node:crypto";

import { AuthClient } from "./authClient";
import { Chunk, RelevantChunk } from "./grpc/chunks";
import { GetDocumentParams, SearchChunksParams, SearchClient } from "./searchClient";

export interface UserData {
  signInState?: string;
  refreshToken?: string;
  idToken?: string;
  idTokenExpiry?: Date;
  connections?: string[];
}

export interface MultiUserSearchChunksParams extends Omit<SearchChunksParams, "accessToken"> {
  userId: string;
}

export interface MultiUserGetDocumentParams extends Omit<GetDocumentParams, "accessToken"> {
  userId: string;
}

export class MultiUserClient {
  authClient: AuthClient;
  searchClient: SearchClient;

  callbackUri: string;

  readUserData: (userId: string) => Promise<UserData | undefined>;
  writeUserData: (userId: string, data: UserData | undefined) => Promise<void>;

  /**
   * Redactive client handling multiple user authentication and access to the Redactive Search service.
   * @param apiKey - Redactive API key.
   * @param callbackUri - The URI to redirect to after initiating the connection.
   * @param readUserData - Function to read user data from storage.
   * @param writeUserData - Function to write user data to storage.
   * @param options - An object of client options. Optional.
   * @param options.authBaseUrl - Base URL for the authentication service. Optional.
   * @param options.grpcHost - Host for the Redactive API service. Optional.
   * @param options.grpcPort - Port for the Redactive API service. Optional.
   */
  constructor(
    apiKey: string,
    callbackUri: string,
    readUserData: (userId: string) => Promise<UserData | undefined>,
    writeUserData: (userId: string, userData: UserData | undefined) => Promise<void>,
    options?: {
      authBaseUrl?: string;
      grpcHost?: string;
      grpcPort?: number;
    }
  ) {
    this.authClient = new AuthClient(apiKey, options?.authBaseUrl);
    this.searchClient = new SearchClient(options?.grpcHost, options?.grpcPort);

    this.callbackUri = callbackUri;

    this.readUserData = readUserData;
    this.writeUserData = writeUserData;
  }

  /**
   * Return a URL for authorizing Redactive to connect with provider on a user's behalf.
   * @param userId - A user ID to associate the connection URL with.
   * @param provider - The name of the provider to connect with.
   * @returns The URL to redirect the user to for beginning the connection.
   */
  async getBeginConnectionUrl(userId: string, provider: string): Promise<string> {
    const state = randomUUID();
    const url = await this.authClient.beginConnection({ provider, redirectUri: this.callbackUri, state });

    const userData = await this.readUserData(userId);
    await this.writeUserData(userId, {
      ...userData,
      signInState: state
    });
    return url;
  }

  _refreshUserData: (userId: string, refreshToken?: string, signInCode?: string) => Promise<UserData> = async (
    userId: string,
    refreshToken?: string,
    signInCode?: string
  ) => {
    const tokens = await this.authClient.exchangeTokens(signInCode, refreshToken);
    const connections = await this.authClient.listConnections(tokens.idToken);
    const userData = {
      refreshToken: tokens.refreshToken,
      idToken: tokens.idToken,
      idTokenExpiry: new Date(Date.now() + (tokens.expiresIn - 10) * 1000),
      connections: connections.connections
    };
    await this.writeUserData(userId, userData);
    return userData;
  };

  async getUsersRedactiveEmail(userId: string): Promise<string | undefined> {
    const userData = await this.readUserData(userId);
    if (!userData || !userData.idToken) {
      return undefined;
    }
    const encodedBody = userData.idToken.split(".")[1];
    const tokenBody = JSON.parse(atob(encodedBody));
    return tokenBody.email;
  }

  /**
   * The callback method for users completing the connection flow; to be called when user returns to app with
   * connection-related URL query parameters.
   * @param userId - The ID of the user completing their connection flow.
   * @param signInCode - The connection sign-in code returned in the URL query parameters by completing the connection flow.
   * @param state - The state value returned in the URL query parameters by completing the connection flow.
   * @returns A boolean representing successful connection completion.
   */
  async handleConnectionCallback(userId: string, signInCode: string, state: string): Promise<boolean> {
    const userData = await this.readUserData(userId);
    if (!userData || !state || userData.signInState !== state) {
      return false;
    }
    await this._refreshUserData(userId, undefined, signInCode);
    return true;
  }

  /**
   * Retrieve the list of the user's provider connections.
   * @param userId - The ID of the user.
   * @returns A list of the user's connected providers.
   */
  async getUserConnections(userId: string): Promise<string[]> {
    let userData = await this.readUserData(userId);
    if (!!userData && !!userData.idTokenExpiry && new Date(userData.idTokenExpiry) > new Date()) {
      return userData.connections ?? [];
    }
    if (!!userData && !!userData.refreshToken) {
      userData = await this._refreshUserData(userId, userData.refreshToken, undefined);
      return userData.connections ?? [];
    }
    return [];
  }

  async clearUserData(userId: string): Promise<void> {
    await this.writeUserData(userId, undefined);
  }

  /**
   * Query for relevant chunks based on a semantic query.
   * @param userId - The ID of the user.
   * @param query - The query string used to find relevant chunks.
   * @param count - The number of relevant chunks to retrieve. Defaults to 10.
   * @param filters - An object of filters for querying. Optional.
   * @returns list of relevant chunks.
   */
  async searchChunks({ userId, query, count = 10, filters }: MultiUserSearchChunksParams): Promise<RelevantChunk[]> {
    let userData = await this.readUserData(userId);
    if (!userData || !userData.refreshToken) {
      throw new Error(`No valid Redactive session for user '${userId}'`);
    }
    if (!!userData.idTokenExpiry && new Date(userData.idTokenExpiry) < new Date()) {
      userData = await this._refreshUserData(userId, userData.refreshToken, undefined);
    }

    return await this.searchClient.searchChunks({ accessToken: userData.idToken!, query, count, filters });
  }

  /**
   * Get chunks from a document by its URL.
   * @param accessToken - The user's Redactive access token.
   * @param ref - A reference to the document we are retrieving. Can be either a url or document name.
   * @param filters - The filters for querying documents. Optional. Only applicable for getting by document name.
   * @returns The complete list of chunks for the matching document.
   */
  async getDocument({ userId, ref, filters }: MultiUserGetDocumentParams): Promise<Chunk[]> {
    let userData = await this.readUserData(userId);
    if (!userData || !userData.refreshToken) {
      throw new Error(`No valid Redactive session for user '${userId}'`);
    }
    if (!!userData.idTokenExpiry && new Date(userData.idTokenExpiry) < new Date()) {
      userData = await this._refreshUserData(userId, userData.refreshToken, undefined);
    }

    return await this.searchClient.getDocument({ accessToken: userData.idToken!, ref, filters });
  }
}

File: sdks/python/src/redactive/multi_user_client.py
import uuid
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from datetime import UTC, datetime, timedelta
from typing import Annotated, Any

import jwt

from redactive.auth_client import AuthClient
from redactive.grpc.v2 import Filters, GetDocumentResponse, SearchChunksResponse
from redactive.search_client import SearchClient


@dataclass
class UserData:
    refresh_token: str | None = None
    id_token: str | None = None
    id_token_expiry: datetime | None = None
    connections: list[str] = field(default_factory=list)
    sign_in_state: str | None = None


class InvalidRedactiveSessionError(Exception):
    def __init__(self, user_id: str) -> None:
        super().__init__(f"No valid Redactive session for user '{user_id}'")


class MultiUserClient:
    def __init__(
        self,
        api_key: str,
        callback_uri: str,
        read_user_data: Callable[[Annotated[str, "user_id"]], Awaitable[UserData]],
        write_user_data: Callable[[Annotated[str, "user_id"], UserData | None], Awaitable[None]],
        *,
        auth_base_url: str | None = None,
        grpc_host: str | None = None,
        grpc_port: int | None = None,
    ) -> None:
        """
        Redactive client handling multiple users authentication and access to the Redactive Search service.

        :param api_key: Redactive API key.
        :type api_key: str
        :param callback_uri: The URI to redirect to after initiating the connection.
        :type callback_uri: str
        :param read_user_data: Function to read user data from storage.
        :type read_user_data: Callable[[Annotated[str, user_id]], Awaitable[UserData]]
        :param write_user_data: Function to write user data to storage.
        :type write_user_data: Callable[[[Annotated[str, user_id], UserData | None], Awaitable[None]]
        :param auth_base_url: Base URL for the authentication service. Optional.
        :type auth_base_url: str | None
        :param grpc_host: Host for the Redactive API service. Optional.
        :type grpc_host: str | None
        :param grpc_port: Port for the Redactive API service. Optional.
        :type grpc_port: int | None
        """

        self.auth_client = AuthClient(api_key, base_url=auth_base_url)
        self.search_client = SearchClient(host=grpc_host, port=grpc_port)
        self.callback_uri = callback_uri
        self.read_user_data = read_user_data
        self.write_user_data = write_user_data

    async def get_begin_connection_url(self, user_id: str, provider: str) -> str:
        """
        Return a URL for authorizing Redactive to connect with provider on a user's behalf.

        :param user_id: A user ID to associate connection URL with.
        :type user_id: str
        :param provider: The name of the provider to connect with.
        :type provider: str
        :return: The URL to redirect the user to for beginning the connection.
        :rtype: str
        """
        state = str(uuid.uuid4())
        response = await self.auth_client.begin_connection(provider, self.callback_uri, state=state)
        user_data = await self.read_user_data(user_id)
        user_data.sign_in_state = state
        await self.write_user_data(user_id, user_data)
        return response.url

    async def _refresh_user_data(
        self, user_id: str, refresh_token: str | None = None, sign_in_code: str | None = None
    ) -> UserData:
        tokens = await self.auth_client.exchange_tokens(sign_in_code, refresh_token)
        connections = await self.auth_client.list_connections(tokens.idToken)
        user_data = UserData(
            refresh_token=tokens.refreshToken,
            id_token=tokens.idToken,
            id_token_expiry=datetime.now(UTC) + timedelta(seconds=tokens.expiresIn - 10),
            connections=connections.connections,
        )
        await self.write_user_data(user_id, user_data)
        return user_data

    async def get_users_redactive_email(self, user_id: str) -> str | None:
        user_data = await self.read_user_data(user_id)
        if not user_data or not user_data.id_token:
            return None
        token_body = jwt.decode(user_data.id_token, options={"verify_signature": False})
        return token_body.get("email")

    async def handle_connection_callback(self, user_id: str, sign_in_code: str, state: str) -> bool:
        """
        The callback method for users completing the connection flow; to be called when user returns to app with
        connection-related URL query parameters.

        :param user_id: The ID of the user completing their connection flow.
        :type user_id: str
        :param sign_in_code: The connection sign-in code returned in the URL query params by completing the connection flow.
        :type sign_in_code: str
        :param state: The state value returned in the URL query params by completing the connection flow.`
        :type state: str
        :return: A boolean represent successful connection completion.
        :rtype: bool
        """
        user_data = await self.read_user_data(user_id)
        if not user_data or user_data.sign_in_state != state:
            return False
        await self._refresh_user_data(user_id, sign_in_code=sign_in_code)
        return True

    async def get_user_connections(self, user_id: str) -> list[str]:
        """
        Retrieve the list of user's provider connections.
        :param user_id: The ID of the user.
        :type user_id: str
        :return: A list of user's connected providers.
        """
        user_data = await self.read_user_data(user_id)
        if user_data and user_data.id_token_expiry and user_data.id_token_expiry > datetime.now(UTC):
            return user_data.connections
        if user_data and user_data.refresh_token:
            user_data = await self._refresh_user_data(user_id, refresh_token=user_data.refresh_token)
            return user_data.connections
        return []

    async def clear_user_data(self, user_id: str) -> None:
        await self.write_user_data(user_id, None)

    async def _get_id_token(self, user_id: str) -> str:
        user_data = await self.read_user_data(user_id)
        if not user_data or not user_data.refresh_token:
            raise InvalidRedactiveSessionError(user_id)
        if user_data.id_token_expiry and user_data.id_token_expiry < datetime.now(UTC):
            user_data = await self._refresh_user_data(user_id, refresh_token=user_data.refresh_token)
        if not user_data.id_token:
            raise InvalidRedactiveSessionError(user_id)
        return user_data.id_token

    async def search_chunks(
        self, user_id: str, query: str, count: int = 10, filters: Filters | dict[str, Any] | None = None
    ) -> SearchChunksResponse:
        """
        Query for relevant chunks based on a semantic query.

        :param user_id: The ID of the user.
        :type user_id: str
        :param query: The query string used to find relevant chunks.
        :type query: str
        :param count: The number of relevant chunks to retrieve. Defaults to 10.
        :type count: int, optional
        :param filters: The filters for relevant chunks. See `Filters` type.
        :type filters: Filters | dict[str, Any], optional
        :return: A list of relevant chunks that match the query
        :rtype: list[RelevantChunk]
        """
        id_token = await self._get_id_token(user_id)
        return await self.search_client.search_chunks(id_token, query, count, filters=filters)

    async def get_document(self, user_id: str, ref: str) -> GetDocumentResponse:
        """
        Get chunks from a document by its URL.

        :param user_id: The ID of the user.
        :type user_id: str
        :param url: The URL to the document for retrieving chunks.
        :type url: str
        :return: The complete list of chunks for the document.
        :rtype: list[Chunk]
        """
        id_token = await self._get_id_token(user_id)
        return await self.search_client.get_document(id_token, ref)
